src/Razor/src/Microsoft.CodeAnalysis.Razor.Workspaces/DefaultProjectWorkspaceStateGenerator.cs

```diff
@@ -162,7 +171,14 @@ private async Task UpdateWorkspaceStateAsync(Project workspaceProject, ProjectSn
                             csharpLanguageVersion = csharpParseOptions.LanguageVersion;
                         }
 
+                        var stopWatch = Stopwatch.StartNew();
                         var tagHelperResolutionResult = await _tagHelperResolver.GetTagHelpersAsync(workspaceProject, projectSnapshot, cancellationToken);
+                        stopWatch.Stop();
+                        _telemetryReporter.ReportEvent("TagHelpersTime", VisualStudio.Telemetry.TelemetrySeverity.Normal, new Dictionary<string, object>()
```

Comment:
I don't think OOP is always a delta, but I think OOP is the only thing that does deltas. Also the deltas, as far as I know, is only about serialization, not discovery, so perhaps separating telemetry for those would make sense too. eg, telemetry for how long it takes to discover tag helpers, how long to serialize/deserialize when not doing a delta, and how long to serialize/deserialize when doing a delta.

I think this really just means moving the telemetry lower down, into the tag helper providers, and the serialization code.

Reply:
Alright, I think this makes more sense. The last bit will be deciding if we need one more thing to track the total time instead of each individual, or if individual is more important.